ui/packages/catalog-ui-search/src/main/webapp/component/metacard-overwrite/metacard-overwrite.view.js

```diff
@@ -13,116 +13,306 @@
+    super(props)
+    this.selectionInterface = props.selectionInterface || store
+    this.state = defaultState
+    this.state.model =
+      props.model || this.selectionInterface.getSelectedResults().first()
+    this.dropzoneElement = React.createRef()
+  }
+
+  componentDidMount() {
+    this.dropzone = new Dropzone(
+      ReactDOM.findDOMNode(this.dropzoneElement.current),
+      {
+        url: './internal/catalog/' + this.state.model.get('metacard').id,
+        maxFilesize: 5000000, //MB
+        method: 'put',
+      }
+    )
     this.trackOverwrite()
     this.setupEventListeners()
     this.handleSending()
     this.handlePercentage()
     this.handleError()
     this.handleSuccess()
-  },
-  trackOverwrite: function() {
-    if (!getOverwriteModel(this)) {
+  }
+
+  render() {
+    const Component = Stages[this.state.stage]
+    return (
+      <Root>
+        <div style={{ display: 'none' }} ref={this.dropzoneElement} />
+        <Component
+          {...this.state}
+          archive={() => this.archive()}
+          startOver={() => this.startOver()}
+        />
+      </Root>
+    )
+  }
+
+  getOverwriteModel() {
+    return OverwritesInstance.get(this.state.model.get('metacard').id)
+  }
+
+  trackOverwrite() {
+    if (!this.getOverwriteModel()) {
       OverwritesInstance.add({
-        id: this.model.get('metacard').id,
+        id: this.state.model.get('metacard').id,
         dropzone: this.dropzone,
-        result: this.model,
+        result: this.state.model,
       })
     }
-  },
-  setupEventListeners: function() {
-    var overwriteModel = getOverwriteModel(this)
-    this.listenTo(overwriteModel, 'change:percentage', this.handlePercentage)
-    this.listenTo(overwriteModel, 'change:sending', this.handleSending)
-    this.listenTo(overwriteModel, 'change:error', this.handleError)
-    this.listenTo(overwriteModel, 'change:success', this.handleSuccess)
-  },
-  handleSending: function() {
-    var sending = getOverwriteModel(this).get('sending')
-    this.$el.toggleClass('show-progress', sending)
-  },
-  handlePercentage: function() {
-    var percentage = getOverwriteModel(this).get('percentage')
-    this.$el
-      .find('.overwrite-progress > .progress-bar')
-      .css('width', percentage + '%')
-    this.$el.find('.progress-percentage').html(Math.floor(percentage) + '%')
-  },
-  handleError: function() {
-    var error = getOverwriteModel(this).get('error')
-    this.$el.toggleClass('has-error', error)
-    this.$el
-      .find('.error-message')
-      .html(getOverwriteModel(this).escape('message'))
-  },
-  handleSuccess: function(file, response) {
-    var success = getOverwriteModel(this).get('success')
-    this.$el.toggleClass('has-success', success)
-    this.$el
-      .find('.success-message')
-      .html(getOverwriteModel(this).escape('message'))
-  },
-  archive: function() {
-    this.listenTo(
+  }
+
+  setupEventListeners() {
+    const overwriteModel = this.getOverwriteModel()
```

Comment:
I think it would be cleaner to use a function that maps your model to the state. Example: https://github.com/codice/ddf/pull/4458/files#diff-0598f5d2aa848d8fec1a22ef1e079296R69

Then you could do something like `listenTo(overwriteModel, 'change:percentage change:sending ...', this.handleChange)`

Reply:
I added this in the latest commit